tokens/cnft-vault/programs/cnft-vault/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+            ctx.accounts.new_leaf_owner1.to_account_info(),
+            ctx.accounts.merkle_tree1.to_account_info(),
+            ctx.accounts.log_wrapper.to_account_info(),
+            ctx.accounts.compression_program.to_account_info(),
+            ctx.accounts.system_program.to_account_info(),
+        ];
+        let mut account_infos2: Vec<AccountInfo> = vec![
+            ctx.accounts.tree_authority2.to_account_info(),
+            ctx.accounts.leaf_owner.to_account_info(),
+            ctx.accounts.leaf_owner.to_account_info(),
+            ctx.accounts.new_leaf_owner2.to_account_info(),
+            ctx.accounts.merkle_tree2.to_account_info(),
+            ctx.accounts.log_wrapper.to_account_info(),
+            ctx.accounts.compression_program.to_account_info(),
+            ctx.accounts.system_program.to_account_info(),
+        ];
+        
+        // add "accounts" (hashes) that make up the merkle proof
+        let mut i = 0u8;
+        for acc in ctx.remaining_accounts.iter() {
+            if i < proof_1_length {
+                accounts1.push(AccountMeta::new_readonly(acc.key(), false));
+                account_infos1.push(acc.to_account_info());
+            } else {
+                accounts2.push(AccountMeta::new_readonly(acc.key(), false));
+                account_infos2.push(acc.to_account_info());
+            }
+            i+=1;
+        }
+
+        msg!("withdrawing cNFT#1");
+        solana_program::program::invoke_signed(
+        & solana_program::instruction::Instruction {
+            program_id: ctx.accounts.bubblegum_program.key(),
+            accounts: accounts1,
+            data: data1,
+        },
+        &account_infos1[..],
+        &[&[b"cNFT-vault", &[*ctx.bumps.get("leaf_owner").unwrap()]]])?;
+        
+        msg!("withdrawing cNFT#2");
+        solana_program::program::invoke_signed(
+        & solana_program::instruction::Instruction {
+            program_id: ctx.accounts.bubblegum_program.key(),
+            accounts: accounts2,
+            data: data2,
+        },
+        &account_infos2[..],
+        &[&[b"cNFT-vault", &[*ctx.bumps.get("leaf_owner").unwrap()]]])?;
+
+        msg!("successfully sent cNFTs");
+        Ok(())
+
+        
+    }
+
+}
+
+#[derive(Accounts)]
+pub struct Withdraw<'info> {
+    #[account(
+        seeds = [merkle_tree.key().as_ref()],
+        bump, 
+        seeds::program = bubblegum_program.key()
+    )]
+    /// CHECK: This account is neither written to nor read from.
+    pub tree_authority: Account<'info, TreeConfig>,
+    
+    #[account(
+        seeds = [b"cNFT-vault"],
+        bump,
+    )]
+    /// CHECK: This account doesnt even exist (it is just the pda to sign)
+    pub leaf_owner: UncheckedAccount<'info>, // sender (the vault in our case)
+    /// CHECK: This account is chekced in the instruction
+    pub leaf_delegate: UncheckedAccount<'info>, // we could actually remove this and just use leaf_owner instead
```

Comment:
Wondering if it would be enough to delegate the xNFT to the program and then revoke the delegation to withdraw.

Reply:
depends what you want to do. and yes, that could be an option 
(cNFT btw)